example-async.js

```diff
@@ -0,0 +1,158 @@
+'use strict'
+
+const Fastify = require('fastify')
+
+function build (opts) {
+  const fastify = Fastify(opts)
+
+  fastify
+    .register(require('fastify-jwt'), { secret: 'supersecret' })
+    .register(require('fastify-leveldb'), { name: 'authdb-async' })
+    .register(require('./fastify-auth'))
+    .after(routes)
+
+  fastify.decorate('verifyJWTandLevel', verifyJWTandLevel)
+  fastify.decorate('verifyUserAndPassword', verifyUserAndPassword)
+
+  function verifyJWTandLevel (request, reply) {
+    const jwt = this.jwt
+    const level = this.level
+
+    return new Promise(function (resolve, reject) {
+      if (request.body && request.body.failureWithReply) {
+        reply.code(401).send({ 'error': 'Unauthorized' })
+        return reject(new Error())
+      }
+
+      if (!request.req.headers['auth']) {
+        return reject(new Error('Missing token header'))
+      }
+
+      jwt.verify(request.req.headers['auth'], onVerify)
```

Comment:
I would use `util.promisify` to get a promise-based version of this.

Reply:
@mcollina can you, please, check if this is what you meant?
https://github.com/PP-etc/fastify-auth/blob/feature/refactor-to-use-util.promisify/example-async.js#L31

To me mix of callback and Promise looks worse, than just returning `Promise` right away and using `resolve/reject` after. 

But I believe I've solved it not in the best possible way, please, point me what can be improved.

Reply:
To me the full `verifyJWTandLevel` function needs to become an async function. 

```js
const decoded = await util.promisify(jwt.verify)(request.req.headers['auth'])
if (!decoded.user || !decoded.password) {
  throw new Error('Token not valid')
}
try {
  const user = await level.get(decoded.user)
 // ...
} catch (err) {
   // ..
}
```

Reply:
I use `Promise` because there are tests for `node 6`: https://travis-ci.org/fastify/fastify-auth/jobs/508269248 .

If it's okay to drop it of course proper `async/await` will look better.

Reply:
You can use promises all the way.

Reply:
https://github.com/PP-etc/fastify-auth/blob/feature/refactor-to-use-util.promisify/example-async.js#L18

please, check if this is what you mean

Reply:
I've replaced `util.promisify` with `new Promise`